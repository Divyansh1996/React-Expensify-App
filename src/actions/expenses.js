import {v4 as uuid4} from 'uuid';
import db from '../firebase/firebase'
import { push,ref,get,remove, update} from 'firebase/database';

export const addExpense = (expense) => (
    {
        type:'ADD_EXPENSE',
        expense
    })

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {description = '', note = '', amount = 0, createdAt = ''} = expenseData
        const expense = {description, note, amount, createdAt}
        return push(ref(db, `users/${uid}/expenses`), expense).then((ref) => {
            dispatch(addExpense({id:ref.key, ...expense}))
        })

    }
}

export const removeExpense=({id=''}={}) => ({
    type:'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense =({id}) => {
    return(dispatch, getState) => {
        const uid = getState().auth.uid;
        return remove(ref(db, `users/${uid}/expenses/${id}`)).then(()=>{
            dispatch(removeExpense({id}))
        })
    }

}


export const editExpense = (id, updates) => ({
    id,
    type:'EDIT_EXPENSE',
    updates
})

export const startEditExpense =(id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return update(ref(db, `users/${uid}/expenses/${id}`),updates).then(() => {
            dispatch(editExpense(id, updates))
        })
    }
}

export const setExpenses = (expenses) => ({
    type: "SET_EXPENSES",
    expenses
})

export const startSetExpenses = () =>{
    return(dispatch, getState) => {
        const uid = getState().auth.uid;
        return get(ref(db, `users/${uid}/expenses`)).then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpenses(expenses))
        })
    }
}