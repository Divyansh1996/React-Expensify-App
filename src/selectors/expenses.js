import moment from "moment";

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const momentCreatedAt = moment(expense.createdAt);
        const startMatch = startDate? startDate.isSameOrBefore(momentCreatedAt):true
        const endMatch = endDate? endDate.isSameOrAfter(momentCreatedAt):true
        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());
        
        return startMatch && endMatch && textMatch;
    }).sort((a,b) => {
        if(sortBy === 'date'){
            return a.createdAt <=b.createdAt?1:-1
        }
        if(sortBy === 'amount'){
            return a.amount <=b.amount?1:-1
        }
    })
}
