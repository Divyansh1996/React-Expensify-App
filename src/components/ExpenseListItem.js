import React from "react";
import { removeExpense } from "../actions/expenses";
import {Link} from "react-router-dom";
import moment from "moment";

const ExpenseListItem = ({id, description, amount, createdAt}) =>(
    <div>
        <Link to={`/edit/${id}`}>{description}</Link>
        <p>{amount} - {moment(createdAt).format('LL')}</p>
    </div>
)

export default ExpenseListItem