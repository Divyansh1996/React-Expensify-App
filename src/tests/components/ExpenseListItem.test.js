import React from "react";
import ExpenseListItem from "../../components/ExpenseListItem"
import expenses from "../fixtures/expenses"
import { shallow } from "enzyme";
import moment from "moment";

test("Testing the ExpenseListItem with few values", () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>)
    expect(wrapper).toMatchSnapshot()
})
