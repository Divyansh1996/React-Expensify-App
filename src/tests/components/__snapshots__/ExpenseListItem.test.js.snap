// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Testing the ExpenseListItem with few values 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    Rent
  </Link>
  <p>
    $1,250.00
     - 
    January 1, 1970
  </p>
</div>
`;
